js/tabRestore.js

```diff
@@ -10,17 +10,35 @@ function TabHistory() {
 }
 
 TabHistory.prototype.push = function (tabId) {
-  var tab = tabs.get(tabId)
+  var closedTab = tabs.get(tabId)
+  var parentTask
+
+  tasks.get().forEach(function (task) {
+    task.tabs.forEach(function (tab) {
+      if (tab.id === tabId) {
+        parentTask = task
+      }
+    })
+  })
+
   // Do not store private tabs or blank tabs
-  if (tab.private || tab.url === 'about:blank' || tab.url === '') {
+  if (closedTab.private
+    || closedTab.url === 'about:blank'
+    || closedTab.url === '') {
     return
   }
 
   if (this.stack.length < this.depth) {
-    this.stack.push(tab)
+    this.stack.push({
+      tab: closedTab,
+      task: parentTask || getSelectedTask(),
```

Comment:
I wrote about this below, too, but you can use directly the `window.currentTask` variable. I'm not really sure why the `getSelectedTask()` function exists at all - perhaps in some edge cases we cannot rely on the global variable? I think there're no such cases as both `switchToTab` and `switchToTask` set it correctly; it's probably left from some old code.
Maybe @PalmerAL has a better explanation?